daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+}
+
+/* Replace the data file with the contents of the given buffer, and update the
+ * metadata file accordingly. In the event of an error, the circular file may
+ * instead be left logically empty, but the data and metadata files will never
+ * be left in an inconsistent state.
+ */
+static gboolean
+overwrite (EmerCircularFile *self,
+           gconstpointer     buffer,
+           gsize             num_bytes,
+           guint64           prev_max_size,
+           GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 little_endian_max_size = swap_bytes_64_if_big_endian (priv->max_size);
+  /* The size of the data file may not exceed the maximum size stored in the
+   * metadata file.
+   */
+  if (priv->max_size > prev_max_size)
+    g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           MAX_SIZE_KEY, little_endian_max_size);
+  if (!set_metadata (self, 0, 0, error))
+    return FALSE;
+
+  gchar *data_filepath = g_file_get_path (priv->data_file);
+  gboolean write_succeeded =
+    g_file_set_contents (data_filepath, buffer, num_bytes, error);
+  g_free (data_filepath);
+  if (!write_succeeded)
+    return FALSE;
+
+  /* We only need to update the maximum size if we didn't already. */
+  if (priv->max_size < prev_max_size)
+    g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           MAX_SIZE_KEY, little_endian_max_size);
+
+  return add_to_size (self, num_bytes, error);
+}
+
+/* Change the maximum size of the circular file from prev_max_size to
+ * priv->max_size. If the new maximum is less than the amount of data currently
+ * in the buffer, then any data that doesn't fit will be removed. Also
+ * reorganizes the circular file so that its head is at the start of the file.
+ */
+static gboolean
+resize (EmerCircularFile *self,
+        guint64           prev_max_size,
+        GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gsize bytes_to_read = MIN (priv->size, priv->max_size);
+  guint8 buffer[bytes_to_read];
+  gboolean read_succeeded =
+    read_disk_bytes (self, buffer, bytes_to_read, prev_max_size, error);
+  if (!read_succeeded)
+    return FALSE;
+
+  guint64 new_size = get_trimmed_size (buffer, bytes_to_read);
+  return overwrite (self, buffer, new_size, prev_max_size, error);
+}
+
+static gboolean
+continue_reading_from_start (EmerCircularFile *self,
+                             GInputStream     *input_stream,
+                             guint8           *partially_filled_buffer,
+                             gsize             total_bytes_to_read,
+                             gsize             bytes_read_end,
+                             GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (bytes_read_end < total_bytes_to_read)
```

Comment:
Could you return early at the beginning of this function when `bytes_read_end >= total_bytes_to_read`?